src/ItemFieldSet.tsx

```diff
@@ -66,9 +117,85 @@ export class ItemFieldSet extends React.Component<IItemFieldSetProps, IItemField
+    const items: IItem[] = [...this.state.items];
+    const newPrice = event.target.value.replace(/[^0-9.]/, '');
+    items[index].price = newPrice;
+    this.setState({ items });
+    this.props.setItems(this.state.items);
+  }
+
   private changeItemQuantity = (index: number) => (event: React.ChangeEvent<HTMLInputElement>) => {
     const items: IItem[] = [...this.state.items];
-    items[index].quantity = parseInt(event.target.value, 10);
+    const newQuantity = event.target.value.replace(/[^0-9]/, '');
+    if (!this.isNumber(newQuantity)) {
+      return
+    }
+    items[index].quantity = newQuantity;
+    this.setState({ items });
+    this.props.setItems(this.state.items);
+  }
+
+  private itemPriceToggleFocus = (index: number, focused: boolean) => (event: React.ChangeEvent<HTMLInputElement>) => {
+    const items: IItem[] = [...this.state.items];
+    items[index].isFocused = focused;
+    const removedDollarSign = items[index].price.replace(/[^\d.-]/g, '');
+    if (removedDollarSign === "") {
+      return;
+    }
+    if (focused) {
+      items[index].price = Money.fromDecimal(parseFloat(removedDollarSign), Currencies.USD, Math.round).toString();
+    } else {
+      items[index].price = "$" + Money.fromDecimal(parseFloat(items[index].price), Currencies.USD, Math.round).toString();
+    }
+    this.setState({ items });
+  }
+
+  private itemQuantityToggleFocus = (index: number, focused: boolean) => (event: React.ChangeEvent<HTMLInputElement>) => {
+    const items: IItem[] = [...this.state.items];
+    items[index].isFocused = focused;
+    if (!focused && items[index].quantity === "") {
+      items[index].quantity = "1"
+    }
+    this.setState({ items });
+  }
+
+  private itemsArrayContainsAtLeastOneItem = () => {
+    return this.state.items[0].name !== "" &&
+      this.state.items[0].quantity !== "" &&
+      this.state.items[0].price !== "" &&
+      this.state.items[0].payingIndexes.length > 0;
+  }
+
+  private isPayingForItemNameToggle = (itemIndex: number, nameIndex: number) => (event: React.ChangeEvent<HTMLInputElement>) => {
+    const items: IItem[] = [...this.state.items];
+    if (items[itemIndex].payingIndexes.indexOf(nameIndex) !== -1) {
+      items[itemIndex].payingIndexes = items[itemIndex].payingIndexes
+        .filter(index => index !== nameIndex)
+    }
+    else {
+      items[itemIndex].payingIndexes.push(nameIndex);
+    }
     this.setState({ items });
   }
+
+  private isPayingForItem = (itemIndex: number, nameIndex: number) => {
+    const item: IItem = this.state.items[itemIndex];
+    return item.payingIndexes.indexOf(nameIndex) !== -1
+  }
+
+  private isNumber = (value: string | number) => {
+    return !isNaN(Number(value.toString()));
+  }
+
+  private isValid = () => {
+    return this.itemsArrayContainsAtLeastOneItem();
+  }
+
+  private setItems = () => {
+    const newItems = [...this.state.items];
+    this.setState({ items: newItems });
```

Comment:
Is setItems called when the person clicks on split bill button? If so, you don't need these two lines (newItems and this.setState), they don't do anything